File: 784-insert-into-a-binary-search-tree/insert-into-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        # ITERATIVE 
        # BASE CASE

        if not root:
            return TreeNode(val)

        cur = root 

        while True:
            if val > cur.val:
                if not cur.right:
                    cur.right = TreeNode(val)
                    break
                cur = cur.right
            else:
                if not cur.left:
                    cur.left = TreeNode(val)
                    break
                cur = cur.left 
        return root 

        # RECURSIVE 
        # BASE CASE 

        if not root:
            return TreeNode(val)

        
        if val < root.val:
            root.left = self.insertIntoBST(root.left , val)

        else:
            root.right = self.insertIntoBST(root.right , val)
        
        return root 
        
        

File: 784-insert-into-a-binary-search-tree/insert-into-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode insertIntoBST(TreeNode root, int val) {
        // ITERATIVE VERSION 
        // BASE CASE 

        if(root == null)
        {
            return new TreeNode(val);
        }

        TreeNode cur = root;

        while(true)
        {
            if(val > cur.val)
            {
                if(cur.right == null)
                {
                    cur.right = new TreeNode(val);
                    break;
                }
                cur = cur.right;
            }
            else
            {
                if(cur.left == null)
                {
                    cur.left = new TreeNode(val);
                    break;
                }
                cur = cur.left; 
            }
        }
        return root;
    }
}
        
        
        //RECURSIVE VERSION 

//         if(root == null)
//         {
//             return new TreeNode(val);
//         }

//         if(val > root.val)
//         {
//             root.right = insertIntoBST( root.right, val);
//         }
//         else
//         {
//             root.left = insertIntoBST(root.left , val);
//         }
//         return root;
        
//     }
// }